wmfdata/hive.py

```diff
@@ -1,40 +1,115 @@
 
-def run(cmds, fmt = "pandas", spark_master='yarn', app_name='wmfdata', spark_config={}):
+def run_cli(commands, format = "pandas", heap_size = 1024, use_nice = True, use_ionice = True):
     """
-    Run one or more Hive queries or command on the Data Lake.
-
-    If multiple commands are provided, only results from the last results-producing command will be returned.
+    Runs SQL commands against the Hive tables in the Data Lake using Hive's command line interface.
+
+    Arguments:
+    * `commands`: the SQL to run. A string for a single command or a list of strings for multiple commands within the same session (useful for things like setting session variables). Passing more than one query is *not* supported, and will usually result in an error.
+    * `format`: what format to return the results in
+        * "pandas": a Pandas data frame
+        * "raw": a TSV string, as returned by the command line interface.
+    * `heap_size`: the amount of memory available to the Hive client. Increase this if a command experiences an out of memory error.
+    * `use_nice`: Run with a lower priority for processor usage.
+    * `use_ionice`: Run with a lower priority for disk access.
     """
 
-    if fmt not in ["pandas", "raw"]:
-        raise ValueError("The `fmt` should be either `pandas` or `raw`.")
-    if type(cmds) == str:
-        cmds = [cmds]
-
-    spark_session = spark.get_session(spark_master, app_name, spark_config)
-
-    result_DF = None
-    for cmd in cmds:
-        cmd_result = spark_session.sql(cmd)
-        
-        # If the result has columns, the command was a query and therefore results-producing.
-        # If not, it was a DDL or DML command and not results-producing.
-        if len(cmd_result.columns) > 0:
-            result_DF = cmd_result
-    if fmt == 'pandas':
-        result = result_DF.toPandas()
-    else:
-        result = result_DF.collect()
+    if type(commands) == str:
+        commands = [commands]
+    if format not in ("pandas", "raw"):
+        raise ValueError("'{}' is not a valid format.".format(format))
+    check_kerberos_auth()
+
+    shell_command = "export HADOOP_HEAPSIZE={0} && "
+    if use_nice:
+        shell_command += "/usr/bin/nice "
+    if use_ionice:
+        shell_command += "/usr/bin/ionice "
+    shell_command += "/usr/bin/hive -S -f {1}"
+
+    result = None
+
+    # Support multiple commands by concatenating them in one file. If the user has passed more than one query,
+    # this will result in a error when Pandas tries to read the resulting concatenated output (unless the queries
+    # happen to produce the same number of columns).
+    #
+    # Ideally, we would return only the last query's results or throw a clearer error ourselves. However,
+    # there's no simple way to determine if multiple queries have been passed or separate their output,
+    # so it's not worth the effort.
+    merged_commands = ";\n".join(commands)
```

Comment:
Out of curiosity, do you ever run multiple queries with a single command? Does anyone else? Wondering if this is "well, it might be useful to someone someday" or if there's concrete use cases motivating the inclusion of this feature. Perhaps it would be better to enforce a "one query, one run" rule and leave multiple-querying for the end-user, or, alternatively, move this functionality into, say, `wmfdata.hive.multirun()` to make it explicit.

Reply:
Multiple queries in the same function call, no. I've never tried to support that. With this function, multiple queries will result in an error. The other run functions won't error, but they also will only return the last set of results so it's not possible to actually use them to run multiple queries. So, in practice, there is a "one run, one query" rule :)

But multiple _statements_, yes. That has mainly been useful for modifying session settings, which wouldn't persist between two function calls (for example, [increasing the map memory setting](https://wikitech.wikimedia.org/wiki/Analytics/Systems/Cluster/Hive/Queries#Hadoop_containers_run_out_of_memory).) That doesn't necessarily apply to every engine (not to Spark, for example), but I try to keep things as consistent as possible between the run functions.

Reply:
Why not enforce that the query string you provide must already have all statements in it? e.g. rather than

```python
query = """
SELECT …
"""
commands = ["SET mapred.job.queue.name=nice", "SET mapreduce.map.memory.mb=4096", "USE wmf", query]
wmfdata.hive.run(commands)
```

it would be

```python
query = """
SET mapred.job.queue.name=nice;
SET mapreduce.map.memory.mb=4096;
USE wmf;
SELECT …
"""
wmfdata.hive.run(query)
```

Reply:
OH. For re-reusing queries in multiple engines. My bad!